File: sort/bubble_sort.py
# ----- Python ---------------------------------------- #
# ----- Sorting Algorithm by Bubble Sort  ------------- #
# ----- Approximate number of operations: 0.75*n*n  --- #
# ----------------------------------------------------- #
import time
import copy
import random
from typing import List


def bubble_sort(list_to_sort: List[int]):
    temp_list = copy.deepcopy(list_to_sort)
    size = len(temp_list)

    for i in range(size - 1):
        for j in range(size - 1, i, -1):
            if temp_list[j] < temp_list[j - 1]:
                key = temp_list[j]
                temp_list[j] = temp_list[j - 1]
                temp_list[j - 1] = key

    return temp_list


if __name__ == "__main__":
    list_to_sort = [random.randint(-10000, 10000) for i in range(2000)]
    print(f"List to sort: {list_to_sort}")

    start = time.time()

    sorted_list = bubble_sort(list_to_sort)

    elapsed = time.time() - start
    print(f"Sorted list: {sorted_list}")
    print(f"Size of input list: {len(list_to_sort)}")
    print(f"Elapsed time: {elapsed} seconds")

File: sort/bubble_sort.cpp
/* ---- C++ ----------------------------------------------
------- Sorting Algorithm by Bubble sort -----------------
------- Approximate number of operations 0.75*n*n --------
------------------------------------------------------- */

#include <iostream>

int bubbleSort(int *arrayToSort, size_t inputSize)
{
    int numOfOperations = 0;
    for (int i = 1; i < inputSize; i++)
    {
        for (int j = inputSize - 1; j > i; j--)
        {
            numOfOperations++; // counting comparisons
            if (arrayToSort[j] < arrayToSort[j - 1])
            {
                int key = arrayToSort[j];
                arrayToSort[j] = arrayToSort[j - 1];
                arrayToSort[j - 1] = key;
                numOfOperations++; // counting swaps
            }
        }
        return numOfOperations;
    }
}

void initializeRandomArray(int *arrayToInitialize, size_t inputSize, int min, int max)
{
    for (size_t i = 0; i < inputSize; i++)
    {
        arrayToInitialize[i] = rand() % (max - min + 1) + min;
    }
}

void printArray(int *arrayToPrint, size_t inputSize)
{
    for (size_t i = 0; i < inputSize; i++)
    {
        std::cout << arrayToPrint[i] << "  ";
    }
    std::cout << std::endl;
}

int main()
{

    int min = -1000;
    int max = 1000;
    const int INPUT_SIZE = 1000;
    int inputArray[INPUT_SIZE];

    initializeRandomArray(inputArray, INPUT_SIZE, min, max);

    std::cout << "The original array is: " << std::endl;
    printArray(inputArray, INPUT_SIZE);

    std::cout << "Sorting process started ..." << std::endl;

    std::cout << "Sorting process finished after " << bubbleSort(inputArray, INPUT_SIZE) << " operations." << std::endl;
    std::cout << "The sorted array is: " << std::endl;
    printArray(inputArray, INPUT_SIZE);

    return 0;
}